gpcontrib/arenadata_toolkit/src/arenadata_toolkit_worker.c

```diff
@@ -0,0 +1,346 @@
+ * our latch to wake it up.
+ */
+static void
+tracking_sighup(SIGNAL_ARGS)
+{
+	int			save_errno = errno;
+
+	got_sighup = true;
+	if (MyProc)
+		SetLatch(&MyProc->procLatch);
+
+	errno = save_errno;
+}
+
+static List*
+get_tracked_dbs()
+{
+	Relation	rel;
+	SysScanDesc scan;
+	HeapTuple	tup;
+	List		*tracked_dbs = NIL;
+	tracked_db_t *trackedDb;
+
+	rel = heap_open(DbRoleSettingRelationId, AccessShareLock);
+	scan = systable_beginscan(rel, InvalidOid, false, NULL, 0, NULL);
+	while (HeapTupleIsValid(tup = systable_getnext(scan)))
+	{
+		bool		isnull;
+		Datum		str_datum;
+		Datum		oid_datum;
+		ArrayType  *a;
+
+		str_datum = heap_getattr(tup, Anum_pg_db_role_setting_setconfig,
+								 RelationGetDescr(rel), &isnull);
+		if (isnull)
+			continue;
+
+		oid_datum = heap_getattr(tup, Anum_pg_db_role_setting_setrole,
+								 RelationGetDescr(rel), &isnull);
+		if (DatumGetObjectId(oid_datum) != InvalidOid)
+			continue;
+
+		oid_datum = heap_getattr(tup, Anum_pg_db_role_setting_setdatabase,
+								 RelationGetDescr(rel), &isnull);
+		if (DatumGetObjectId(oid_datum) == InvalidOid)
+			continue;
+
+		a = DatumGetArrayTypeP(str_datum);
+
+		if (is_db_tracked(a))
+		{
+			trackedDb = (tracked_db_t *) palloc0(sizeof(tracked_db_t));
+
+			trackedDb->dbid = DatumGetObjectId(oid_datum);
+			trackedDb->get_full_snapshot_on_recovery = full_snapshot_on_recovery(a);
+			tracked_dbs = lappend(tracked_dbs, trackedDb);
+		}
+	}
+
+	systable_endscan(scan);
+	heap_close(rel, AccessShareLock);
+
+	return tracked_dbs;
+}
+
+static void
+track_dbs(List *tracked_dbs)
+{
+	ListCell   *cell;
+	tracked_db_t *trackedDb;
+
+	foreach(cell, tracked_dbs)
+	{
+		trackedDb = (tracked_db_t *) lfirst(cell);
+
+		bloom_set_bind(&tf_shared_state->bloom_set, trackedDb->dbid);
+		bloom_set_trigger_bits(&tf_shared_state->bloom_set, trackedDb->dbid,
+							   trackedDb->get_full_snapshot_on_recovery);
+	}
+}
+
+/* scan pg_db_role_setting, find all databases, bind blooms if necessary */
+void
+arenadata_toolkit_main(Datum main_arg)
+{
+	elog(LOG, "[arenadata toolkit] Starting background worker");
+
+	/*
+	 * The worker shouldn't exist when the master boots in utility mode.
+	 * Otherwise BackgroundWorkerInitializeConnection will explode with FATAL.
+	 */
+	if(IS_QUERY_DISPATCHER() && Gp_role != GP_ROLE_DISPATCH)
+	{
+		proc_exit(0);
+	}
+
+	/*
+	 * Kludge for scanning pg_db_role_setting on segments.
+	 */
+	if (!IS_QUERY_DISPATCHER() && Gp_role == GP_ROLE_DISPATCH)
+	{
+		Gp_role = GP_ROLE_UTILITY;
+		Gp_session_role = GP_ROLE_UTILITY;
+	}
+
+	pqsignal(SIGHUP, tracking_sighup);
+	pqsignal(SIGTERM, tracking_sigterm);
+
+	BackgroundWorkerUnblockSignals();
+
+	BackgroundWorkerInitializeConnection(DB_FOR_COMMON_ACCESS, NULL);
+
+	while (!got_sigterm)
+	{
+		int			rc;
+		List	   *tracked_dbs = NIL;
+
+		StartTransactionCommand();
+		tracked_dbs = get_tracked_dbs();
+
+		if (!tf_shared_state->is_initialized && list_length(tracked_dbs) > 0)
```

Comment:
Any use (reading/writing) of `tf_shared_state` must be done under an exclusive lock!

Reply:
Since we have lock granularity (separate lock for state variables, separate lock for bloom_set, separeate lock for each entry) I see no reason put the exclusive lock everywhere. In this particular example the tf_shared_state is written only by worker process, others are supposed to be reades

Reply:
why not use `pg_atomic_flag` for this?